### DISCUSSION TITLE
Need help in understanding async operations on atom

### DISCUSSION THREAD

**Discussion by kittssh99** [pmndrs/jotai#1021] [Q&A] [answered]
We recently started moving from react context to Jotai for state management in our existing react project. Currently we have multiple hooks handling different business logic which are common for multiple screens/components.

**Problem :** 
We have an atom having string value, we are trying to update its value which is set by promise resolve. Also this is happening inside a hook which is used by multiple other hooks. We have observed that as soon as the atom’s setter gets called, all the dependent hooks including the current hook starts re-rendering multiple times while the then-block is still executing which is causing slowness in the app.

We also noticed that this problem gets resolved once we move the setter block outside the promise. We replaced Jotai with context and didn't face this issue with the same codebase.

**We have created a small sandbox to replicate the issue:**

https://codesandbox.io/s/fervent-water-v3bwhz
**PS**: Kindly look into the console to track rendering issue.

Any help in understanding this behaviour or suggestions would be really appreciated.

**Comment by dai-shi** [comment, marked as answer]
A quick confirmation: If you put all console.log statements in useEffect(), is the behavior expected? Or, do you still observe something unexpected?

**Reply by kittssh99** [reply]
Hi @dai-shi thank you for the quick reply. We are actually expecting less renders. If you look closely at the logs we have one of the render happening with old state of atom from the hook. What we observed that renders happens even before the promise block has completed execution.

**Reply by dai-shi** [reply]
Okay, I had a closer look, and got what you mean.
This is an unfortunate behavior that can't be easily fixed from the library perspective.
It will be fixed in React 18 with automatic batching.
Meanwhile, if this is a serious problem in apps, one could work around with `unstable_batchedUpdates`.
https://codesandbox.io/s/pensive-snyder-647wxo?file=/src/hooks/useDummyValue.js

**Reply by kittssh99** [reply]
Hi @dai-shi thank you so much, this is exactly what we needed! We used your suggestion in our app which is using react 17.0.2 version and it is working as expected. Just to add for future reference if anyone is using it in react-native app, `unstable_batchedUpdates` are available inside react-native package.
